rcljava/src/main/java/org/ros2/rcljava/time/TimeSource.java

```diff
@@ -0,0 +1,161 @@
+        clock.setRosTimeOverride(timeFromMsg);
+      }
+    }
+  }
+
+  public void setRosTimeIsActive(boolean enabled) {
+    if (this.rosTimeIsActive == enabled) {
+      return;
+    }
+    this.rosTimeIsActive = enabled;
+    for (Clock clock : this.associatedClocks) {
+      clock.setRosTimeIsActive(enabled);
+    }
+    if (enabled) {
+      if (this.node != null) {
+        this.clockSub = node.<rosgraph_msgs.msg.Clock>createSubscription(rosgraph_msgs.msg.Clock.class, "/clock", new SubscriptionCallback(this));
+      }
+    } else {
+      if (this.node != null && this.clockSub != null) {
+        this.node.removeSubscription(this.clockSub);
+        this.clockSub = null;
+      }
+    }
+  }
+
+  public void attachNode(Node node) {
+    if (this.node != null) {
+      detachNode();
+    }
+
+    this.node = node;
+
+    if (!this.node.hasParameter("use_sim_time")) {
+      this.node.declareParameter(new ParameterVariant("use_sim_time", false));
+    }
+
+    ParameterVariant useSimTime = this.node.getParameter("use_sim_time");
+    ParameterType useSimTimeType = useSimTime.getType();
+    if (useSimTimeType != ParameterType.PARAMETER_NOT_SET) {
+      if (useSimTimeType == ParameterType.PARAMETER_BOOL) {
+        this.rosTimeIsActive = useSimTime.asBool();
+      }
+      // TODO(clalancette): what to do if it is the wrong type?
+    }
+
+    class SimTimeCB implements ParameterCallback {
+      private TimeSource timeSource;
+
+      public SimTimeCB(TimeSource ts) {
+        this.timeSource = ts;
+      }
+
+      public rcl_interfaces.msg.SetParametersResult callback(List<ParameterVariant> parameters) {
+        rcl_interfaces.msg.SetParametersResult result = new rcl_interfaces.msg.SetParametersResult();
+        result.setSuccessful(true);
+        for (ParameterVariant param : parameters) {
+          if (param.getName() == "use_sim_time") {
+            if (param.getType() == ParameterType.PARAMETER_BOOL) {
+              this.timeSource.rosTimeIsActive = param.asBool();
+            } else {
+              result.setSuccessful(false);
+              result.setReason("'use_sim_time' parameter must be a boolean");
+              break;
+            }
+          }
+        }
+        return result;
+      }
+    }
+
+    this.simTimeCB = new SimTimeCB(this);
+
+    this.node.addOnSetParametersCallback(this.simTimeCB);
+  }
+
+  public void detachNode() {
+    if (this.node == null) {
+      return;
+    }
+    this.setRosTimeIsActive(false);
+    this.node.removeOnSetParametersCallback(this.simTimeCB);
+    this.node = null;
+  }
+
+  public void attachClock(Clock clock) {
+    if (clock.getClockType() != ClockType.ROS_TIME) {
+      // TODO(clalancette): Make this a custom exception
+      throw new IllegalArgumentException("Cannot attach a clock that is not ROS_TIME to a timesource");
+    }
+    clock.setRosTimeOverride(this.lastTimeSet);
+    clock.setRosTimeIsActive(this.rosTimeIsActive);
+    this.associatedClocks.add(clock);
```

Comment:
what if clocks are attached from multiple threads? `ArrayList` isn't thread safe.

Reply:
Yeah, good call.  I've switched to a `LinkedBlockingQueue` in 63bf582 .